File: backend/app/routers/gift.py
"""
用户端礼品路由 - 礼品列表 / 礼品详情
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.dependencies import get_db, get_current_user
from app.models.user import User
from app.models.gift import Gift

router = APIRouter(prefix="/api/gifts", tags=["礼品"])


@router.get("")
def list_gifts(
    category: str = Query(None, description="分类筛选"),
    db: Session = Depends(get_db),
):
    """用户端：只返回上架礼品，支持分类筛选"""
    query = db.query(Gift).filter(Gift.status == "上架")
    if category:
        query = query.filter(Gift.category == category)

    gifts = query.order_by(Gift.create_time.desc()).all()

    items = []
    for g in gifts:
        items.append({
            "id": g.id,
            "name": g.name,
            "description": g.description,
            "image": g.image,
            "points_required": g.points_required,
            "price": float(g.price) if g.price else 0.00,
            "stock": g.stock,
            "category": g.category,
            "status": g.status,
            "create_time": g.create_time.isoformat() if g.create_time else None,
        })

    return {"code": 200, "data": {"items": items}}


@router.get("/{gift_id}")
def get_gift_detail(
    gift_id: int,
    db: Session = Depends(get_db),
):
    """用户端：礼品详情"""
    g = db.query(Gift).filter(Gift.id == gift_id, Gift.status == "上架").first()
    if not g:
        raise HTTPException(status_code=404, detail="礼品不存在或已下架")

    return {
        "code": 200,
        "data": {
            "id": g.id,
            "name": g.name,
            "description": g.description,
            "image": g.image,
            "points_required": g.points_required,
            "price": float(g.price) if g.price else 0.00,
            "stock": g.stock,
            "category": g.category,
            "status": g.status,
            "create_time": g.create_time.isoformat() if g.create_time else None,
        },
    }

File: frontend/user-app/src/api/gift.ts
/**
 * 礼品 API
 */
import request from "./request";

export interface GiftItem {
  id: number;
  name: string;
  description: string;
  image: string;
  points_required: number;
  price: number;
  stock: number;
  category: string;
  status: string;
  create_time: string;
}

export const giftApi = {
  /** 用户端：礼品列表 */
  list(category?: string) {
    return request.get<{ code: number; data: { items: GiftItem[] } }>("/api/gifts", {
      params: category ? { category } : {},
    });
  },

  /** 用户端：礼品详情 */
  detail(giftId: number) {
    return request.get<{ code: number; data: GiftItem }>(`/api/gifts/${giftId}`);
  },
};

/** 管理端礼品表单 */
export interface GiftForm {
  name: string;
  description: string;
  image: string;
  points_required: number;
  price: number;
  stock: number;
  category: string;
  status: string;
}

/** 管理端礼品列表响应 */
export interface GiftListResponse {
  code: number;
  data: { items: GiftItem[] };
}

export const giftAdminApi = {
  /** 管理端：礼品列表（含库存、状态） */
  list() {
    return request.get<GiftListResponse>("/api/admin/gifts");
  },

  /** 管理端：创建礼品 */
  create(data: GiftForm) {
    return request.post<{ code: number; data: GiftItem }>("/api/admin/gifts", data);
  },

  /** 管理端：编辑礼品 */
  update(giftId: number, data: GiftForm) {
    return request.put<{ code: number; data: GiftItem }>(`/api/admin/gifts/${giftId}`, data);
  },

  /** 管理端：删除礼品 */
  delete(giftId: number) {
    return request.delete<{ code: number }>(`/api/admin/gifts/${giftId}`);
  },
};
